contracts/ZkDai.sol

```diff
@@ -0,0 +1,153 @@
+pragma solidity ^0.4.25;
+
+import 'openzeppelin-solidity/contracts/token/ERC20/ERC20.sol';
+import "./MintNotes.sol";
+import "./SpendNotes.sol";
+
+contract ZkDai is MintNotes, SpendNotes {
+  ERC20 internal DAI_TOKEN_ADDRESS;
+
+  constructor(uint256 _cooldown, uint256 _stake, address daiTokenAddress)
+    public {
+      cooldown = _cooldown;
+      stake = _stake;
+      DAI_TOKEN_ADDRESS = ERC20(daiTokenAddress);
+  }
+
+  /**
+  * @dev Transfers specified number of dai tokens to itself and submits the zkSnark proof to mint a new note
+  * @notice params: a, a_p, b, b_p, c, c_p, h, k zkSnark parameters
+  * @param input Public inputs of the zkSnark
+  */
+  function mint(
+      uint[2] a,
+      uint[2] a_p,
+      uint[2][2] b,
+      uint[2] b_p,
+      uint[2] c,
+      uint[2] c_p,
+      uint[2] h,
+      uint[2] k,
+      uint[4] input)
+    external
+    payable {
+      // check that the first note (among public params) is not already minted 
+      bytes32 note = calcNoteHash(input[0], input[1]);
+      require(notes[note] == State.Invalid, 'Note was already minted');
+      require(msg.value == stake, 'Invalid stake amount');
```

Comment:
Yes, I considered it.
For `mint`, the contract transfers the dai from user to itself. I thought about having this dai as stake until the note is minted. But this gives the user an opportunity to cheat with a little downside - I could say I want to mint a note of value 1 dai but could send in the proof of 1M dai.

If the stake amount is fixed in dai, it would generally mean higher gas requirements (~56k) to transfer the stake back to submitter/challenger as opposed to 21k for vanilla ether transfer.

Reply:
Yeah, i think ETH is fine, but if Dai is used we should do extra Dai that doesn't go in the note

Reply:
Yes.